### ISSUE TITLE
No such file or directory:crypto/conf/conf_def.c:773:calling stat(fipsmodule.cnf)

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#14899]
When selective debugging specific tests that involve loading a provider via `test/fips-and-base.cnf` or the like, e.g.,
```
test/evp_libctx_test -config test/fips-and-base.cnf -provider fips
```
I often run into the annoying situation that I get
```
Error loading config from file test/fips-and-base.cnf
4087E970517F0000:error:80000002:system library:process_include:No such file or directory:crypto/conf/conf_def.c:773:calling stat(fipsmodule.cnf)
4087E970517F0000:error:07800069:common libcrypto routines:provider_conf_load:provider section error:crypto/provider_conf.c:90:section=fips_sect not found
4087E970517F0000:error:0700006D:configuration file routines:module_run:module initialization error:crypto/conf/conf_mod.c:242:module=providers, value=provider_sect retcode=-1      

```
because the relative addressing `.include fipsmodule.cnf` in that config file goes wrong.

As `doc/man5/config.pod` states:
```
As a general rule, the B<pathname> should be an absolute path.  Relative
paths are evaluated based on the current working directory, so unless the
file with the B<.include> directive is application-specific, the inclusion
will not work as expected.  The environment variable B<OPENSSL_CONF_INCLUDE>,
if it exists, will be prepended to all B<.include> B<pathname>'s.
```

A workaround is to use, e.g., `OPENSSL_CONF_INCLUDE=./providers`.

IMO it is bad design to interpret the `.include` pathname relative to the current directory of the running application,
which makes the interpretation context-dependent and thus very fragile.
I suggest interpreting it instead relative to the directory containing the given config file (which is in this case `test/`).

**Comment by DDvO** [comment]
With the suggested improvement of the `.include` semantics, 
```
.include fipsmodule.cnf
```
in `test/fips-and-base.cnf` will become
```
.include ../providers/fipsmodule.cnf
```

**Comment by t8m** [comment]
Unfortunately it is not really possible to get the directory from the file. You'd have to track that somehow from the top-level config file opened and even that is quite problematic. If it was that easy it would be done already. It is documented to behave like this, so this is definitely not a bug but RFE.

Also as the existing semantics is already there and documented, we cannot change the semantics unconditionally.

**Comment by DDvO** [comment]
> Unfortunately it is not really possible to get the directory from the file. You'd have to track that somehow from the top-level config file opened and even that is quite problematic. If it was that easy it would be done already.

Hmm, isn't it possible to derive the directory of an opened file, e.g., from its handle?

> It is documented to behave like this, so this is definitely not a bug but RFE.

Good point.

**Comment by DDvO** [comment]
> Also as the existing semantics is already there and documented, we cannot change the semantics unconditionally.

Ah, yes, the `.include` feature was introduced already in OpenSSL 1.1.1-pre3-dev, in commit b524b808a1d1ba204dbdcbb42de4e3bddb3472ac.

So when the suggested improved semantics is implemented, it would likely be better to rename the feature, maybe to `#include` or the like, and to deprecate the use of `.include`.
This way we could also get rid of the legacy `=` char, which is currently optional.

**Comment by t8m** [comment]
> Hmm, isn't it possible to derive the directory of an opened file, e.g., from its handle?
> 

No, it isn't. At least not on the POSIX systems - otherwise hardlinks would not be possible.

**Comment by DDvO** [comment]
I see. Another possibility should be this string-based approach:
The code that opens e.g, `test/fips-and-base.cnf` knows this pathname.
It can chop the trailing file name and append instead the relative pathname of the file to be included, in this case `../providers/fipsmodule.cnf`, which yields `test/../providers/fipsmodule.cnf`, which should lead to the right location `./providers/fipsmodule.cnf`.

**Comment by mattcaswell** [comment]
OTC: This won't be changed for 3.0.

**Comment by t8m** [comment]
#15090 that was merged has some possible way to solve this - We could pre-process the conf files during build to add the pragma to set the $(OPENSSLDIR) as the includedir.

**Comment by nhorman** [comment]
Can't OPENSSL_CONF_INCLUDE be used to fix this?  I.e. just set that environment variable prior to running tests that use the pragma?
